From aad2339bfe7abd80202881e30f331ff2671bbf39 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Wed, 21 Nov 2018 18:49:23 +0100
Subject: [PATCH] TRITON-716 Add bhyve to list of allowed brands to do
 snapshots Reviewed by: Mike Gerdts <mike.gerdts@joyent.com> Reviewed by:
 Marsell Kukuljevic <marsell@joyent.com> Approved by: Marsell Kukuljevic
 <marsell@joyent.com>

---
 lib/endpoints/vms.js | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/lib/endpoints/vms.js b/lib/endpoints/vms.js
index 1f24c5a..3d37dff 100644
--- a/lib/endpoints/vms.js
+++ b/lib/endpoints/vms.js
@@ -24,6 +24,9 @@ var common = require('../common');
 var errors = require('../errors');
 var interceptors = require('../interceptors');
 
+// First Platform Image supporting Bhyve VMs snapshots:
+const MIN_BHYVE_SNAPSHOT_PLATFORM = '20181119T131511Z';
+
 var VALID_VM_ACTIONS = [
     'start',
     'stop',
@@ -1472,7 +1475,7 @@ function createVm(req, res, next) {
  */
 function canSnapshot(vm) {
 
-    if (['joyent', 'joyent-minimal', 'lx'].indexOf(vm.brand) === -1) {
+    if (['joyent', 'joyent-minimal', 'lx', 'bhyve'].indexOf(vm.brand) === -1) {
         return (new errors.BrandNotSupportedError(
             'snapshots are not supported for VMs of brand "' + vm.brand + '"'));
     }
@@ -1495,6 +1498,17 @@ function canSnapshot(vm) {
             + 'supported for docker VMs that use --volumes-from'));
     }
 
+    /*
+     * Not all the Platform Images will support bhyve snapshots. Let's fail
+     * early instead of create a Job which will fail anyway
+     */
+    if (vm.brand === 'bhyve' &&
+        vm.platform_buildstamp < MIN_BHYVE_SNAPSHOT_PLATFORM) {
+        return (new errors.BrandNotSupportedError('snapshots are not '
+            + 'supported for bhyve VMs unless Platform Image is "' +
+            MIN_BHYVE_SNAPSHOT_PLATFORM + '" or newest'));
+    }
+
     return true;
 }
 
-- 
2.21.0

